lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1993,13 +2034,13 @@ class HSAQueue final : public KalmarQueue
 
         // create shared_ptr instance
         std::shared_ptr<HSABarrier> barrier = std::make_shared<HSABarrier>(this, 0, nullptr);
-        // associate the barrier with this queue
-        pushAsyncOp(barrier);
 
         // enqueue the barrier
         status = barrier.get()->enqueueAsync(release_scope);
         STATUS_CHECK(status, __LINE__);
 
+        // associate the barrier with this queue
+        pushAsyncOp(barrier);
```

Comment:
Would it make sense to move all these `pushAsyncOp` into the various `enqueueAsync` functions (where the queue is locked) such that the copy of the AQL packet into the user mode queue (and among other things) and the cache of the async op are atomic?

Reply:
Yes!  Agree completely.

Reply:
`hcc_memory_async_copy` and `hcc_memory_async_copy_rect` never lock the qmutex -- when is it correct to call `pushAsyncOp` in those cases?

Reply:
I think qmutex should be locked on entrance of these 2 functions because the analysis of the sync optimization needs protection.  Also, the call to the hsa async copy function and the push of the async op into the vector need to be atomic.